server/plugins/messageStorage/sqlite.ts

```diff
@@ -198,6 +251,116 @@ class SqliteMessageStorage implements SearchableMessageStorage {
 		});
 	}
 
+	async fetch_rollbacks(since_version: number) {
+		const res = await this.serialize_fetchall(
+			`select version, rollback_forbidden, statement
+			from rollback_steps
+			join migrations on migrations.id=rollback_steps.migration_id
+			where version > ?
+			order by version desc, step asc`,
+			since_version
+		);
+		const result: Rollback[] = [];
+
+		// convert to Rollback[]
+		// requires ordering in the sql statement
+		res.reduce((combined: Rollback[], cur) => {
+			const last = combined.at(-1);
+
+			if (!last || cur.version !== last.version) {
+				combined.push({
+					version: cur.version,
+					rollback_forbidden: Boolean(cur.rollback_forbidden),
+					stmts: [cur.statement],
+				});
+			} else {
+				last.stmts.push(cur.statment);
+			}
+
+			return combined;
+		}, result);
+		return result;
+	}
+
+	async delete_migrations_older_than(version: number) {
+		return this.serialize_run("delete from migrations where migrations.version > ?", [version]);
+	}
+
+	async _downgrade_to(version: number) {
+		const _rollbacks = await this.fetch_rollbacks(version);
+
+		if (_rollbacks.length === 0) {
+			return version;
+		}
+
+		const forbidden = _rollbacks.find((item) => item.rollback_forbidden);
+
+		if (forbidden) {
+			throw Error(`can't downgrade past ${forbidden.version}`);
+		}
+
+		for (const rollback of _rollbacks) {
+			for (const stmt of rollback.stmts) {
+				await this.serialize_run(stmt, []);
+			}
+		}
+
+		await this.delete_migrations_older_than(version);
+		await this.update_version_in_db();
+
+		return _rollbacks.at(-1)!.version; // assert valid due to length guard above
+	}
+
+	async downgrade_to(version: number) {
+		if (version <= 0) {
+			throw Error(`${version} is not a valid version to downgrade to`);
+		}
+
+		await this.serialize_run("BEGIN EXCLUSIVE TRANSACTION", []);
+
+		let new_version: number;
+
+		try {
+			new_version = await this._downgrade_to(version);
+		} catch (err) {
+			await this.serialize_run("ROLLBACK", []);
+			throw err;
+		}
+
+		await this.serialize_run("COMMIT", []);
+		return new_version;
+	}
+
+	async downgrade() {
+		return this.downgrade_to(currentSchemaVersion);
```

Comment:
I think this and `return await this.downgrade_to(currentSchemaVersion)` is equivalent thanks to the fact that returning a promise from an async function implicitly chains that promise with the parent function's promise. But it strikes me as one of the more esoteric facts about JS and I'd think making `downgrade` a non-async function or awaiting `downgrade_to`'s promise would be clearer.

Reply:
ok, I'll await the result then... don't really want to switch to promises if I don't have to